vidarr-server/src/main/java/ca/on/oicr/gsi/vidarr/server/PriorityByWorkflow.java

```diff
@@ -0,0 +1,147 @@
+              "The current number of workflows that are on priority waiting to run.")
+          .labelNames("workflow")
+          .register();
+
+  private final Map<String, WaitingState> workflows = new ConcurrentHashMap<>();
+
+  @Override
+  public Optional<Pair<String, BasicType>> inputFromSubmitter() {
+    return Optional.of(new Pair<String, BasicType>("Priority", BasicType.INTEGER));
+  }
+
+  @Override
+  public void recover(String workflowName, String workflowVersion, String vidarrId, Optional<JsonNode> resourceJson) {
+    final var stateWaiting = workflows.computeIfAbsent(workflowName, k -> new WaitingState()).waiting;
+    // since we just created it if it doesn't exist, no need for null check here
+
+    int workflowPriority = 1;
+
+    if (resourceJson.isPresent()) {
+      workflowPriority = resourceJson.get().asInt();
+    }
+
+    stateWaiting.add(new SimpleEntry(vidarrId, workflowPriority));
+    currentInPriorityWaitingCount.labels(workflowName).set(stateWaiting.size());
+  }
+
+  @Override
+  public void release(String workflowName, String workflowVersion, String vidarrId) {
+    final var state = workflows.get(workflowName);
+    if (state != null) {
+      state.waiting.remove(vidarrId);
+      currentInPriorityWaitingCount.labels(workflowName).set(state.waiting.size());
+    }
+  }
+
+  @Override
+  public synchronized ConsumableResourceResponse request(
+      String workflowName, String workflowVersion, String vidarrId, Optional<JsonNode> input) {
+
+    int workflowPriority = 1;
+    if (!input.isEmpty()) {
+      workflowPriority = input.get().asInt();
+    }
+
+    if (!acceptedPriorities.contains(workflowPriority)){
+      return ConsumableResourceResponse.error(
+          String.format("Vidarr error: The workflow %s run priority is invalid.", workflowName));
+    }
+    final var state = workflows.get(workflowName);
+    if (state == null) {
+      return ConsumableResourceResponse.error(
+          String.format("Internal Vidarr error: The %s workflow run priority has not been configured properly.", workflowName));
+    } else {
+      if (workflowPriority >= state.waiting.last().getValue()) {
+        state.waiting.remove(new SimpleEntry(vidarrId, workflowPriority));
+        currentInPriorityWaitingCount.labels(workflowName).set(state.waiting.size());
+        return ConsumableResourceResponse.AVAILABLE;
+      } else {
+        state.waiting.add(new SimpleEntry(vidarrId, workflowPriority));
+        currentInPriorityWaitingCount.labels(workflowName).set(state.waiting.size());
+        return ConsumableResourceResponse.error(
+            String.format("There are %s workflows currently queued up with higher priority.", workflowName));
+      }
+    }
+  }
+
+  public void set(String workflowName, String vidarrId, JsonNode input) {
+
+    int workflowPriority = 1;
+    if (!input.isEmpty() && !input.get("Priority").isEmpty()) {
+      workflowPriority = input.get("Priority").asInt();
+    }
```

Comment:
```suggestion
    if (input != null && !input.isEmpty() && input.get("priority") != null && !input.get("priority").isEmpty()) {
      workflowPriority = input.get("priority").asInt();
    }
```
This one isn't an `Optional`, so it could be null (and/or empty).

And this is following up on the earlier request to change `"Priority"` to `"priority"`, as camelCase is idiomatic for JSON, which is what will be received by the `/submit` endpoint.